### ISSUE TITLE
[Question] Randomised train/dev/test split for cross-validation

### ISSUE THREAD

**Issue by lnalborczyk** [espnet/espnet#2955]
Hi,

I am working with a clone of the ljspeech ASR recipe (egs/ljspeech/asr1). I would like to randomly split the available data in train/dev/test sets. In the `./run.sh` of this recipe, the dev set is currently made as follows:

https://github.com/espnet/espnet/blob/fcf20cc33acd09b8ae80c88703aee0b755ed7652/egs/ljspeech/asr1/run.sh#L90-L95

Where `./utils/subset_data_dir.sh` is used to pick the `--first` or `--last` 250 sentences in the `deveval` folder. In other words, the split is not random (it is always the same).

In order to cross-validate the evaluation of my network, I would like to randomise the train/dev/test split (i.e., so that different items are found in the training/test sets at each run). I have tried not using the `--first` or `--last` option of `./utils/subset_data_dir.sh`, but then if I do:

```
# make a dev set 
 utils/subset_data_dir.sh data/${trans_type}_train 500 data/${trans_type}_deveval 
 utils/subset_data_dir.sh --last data/${trans_type}_deveval 250 data/${eval_set} 
 utils/subset_data_dir.sh --first data/${trans_type}_deveval 250 data/${dev_set}
 n=$(( $(wc -l < data/${trans_type}_train/wav.scp) - 500 )) 
 utils/subset_data_dir.sh data/${trans_type}_train ${n} data/${train_set} 
```

Some items may be found both in training and dev/test sets...

Is there an easy way of randomising the train/dev/test split and ensuring that the training/test sets are disjoint in Kaldi/ESPnet?

**Comment by sw005320** [comment]
Probably you can find a better way, but how about making a shuffled list by yourself (e.g., using https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/shuffle_list.pl)?
Then, by using `utility/subset_data_dir.sh --utt-list` (https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/subset_data_dir.sh#L30), you can shuffle the list more explicitly.

**Comment by lnalborczyk** [comment]
Oh great, thank you! I did not notice the `--utt-list` option in `utils/subset_data_dir.sh`...

**Comment by sw005320** [comment]
Year, this script provides several other useful options https://github.com/kaldi-asr/kaldi/blob/master/egs/wsj/s5/utils/subset_data_dir.sh#L15-L32
I think you may find a better way.

**Comment by lnalborczyk** [comment]
Thanks! Here is my (inelegant) implementation, if it can be useful for others.

    # make a dev set
    utils/copy_data_dir.sh data/${trans_type}_train data/${trans_type}_train_full
    rm -rf data/${trans_type}_train

    # make a random list of items
    perl utils/shuffle_list.pl ${db_root}/data_p5/text > shuffled_list
    head -81 shuffled_list > shuffled_train
    tail -40 shuffled_list > shuffled_test

    # splits data intro training and validation/testing sets
    utils/subset_data_dir.sh --utt-list shuffled_train data/${trans_type}_train_full data/${train_set}
    utils/subset_data_dir.sh --utt-list shuffled_test data/${trans_type}_train_full data/${trans_type}_deveval

    # remove temp lists
    rm shuffled_list shuffled_train shuffled_test

    # splits non-training data into validation and testing
    utils/subset_data_dir.sh --last data/${trans_type}_deveval 20 data/${eval_set}
    utils/subset_data_dir.sh --first data/${trans_type}_deveval 20 data/${dev_set}

NB: I only have 121 sentences listed in text, which are subsequently splitted in 81, 20, and 20 sentences for training, validation, and testing, respectively.

I think the issue can be closed.

**Comment by sw005320** [comment]
Very cool.
Thanks!

> EDIT: Actually, I recommend using `cat ${db_root}/data_p5/text | shuf > shuffled_list` instead of perl `utils/shuffle_list.pl ${db_root}/data_p5/text > shuffled_list` as `shuffle_list.pl` shuffles the list in a deterministic way (the same order is applied at each execution).

I think this deterministic way is intentional to get the same result for every experimental trial. `utils/shuffle_list.pl` accepts a random seed number `--srand`, and you can change this number to get different shuffled results but still we can keep the reproducibility if we use the same set of the seed numbers.
